### PR TITLE
Batchnorm training tensor

### PR DESCRIPTION
**Pull request by BigRedT**
Batch norm now works with boolean tensor to indicate training vs testing mode. This allows evaluation in the same script as the training with the same graph!

### CODE REVIEW

File: aiutils/tftools/test_batch_normalizer.py [hunk 1]
```diff
@@ -85,3 +85,86 @@ def test_batch_normalize_fcn(graph):
...
+        ema_mean_eval1 = ema_mean.eval()
+
+        y_eval2 = y.eval(feed_dict={x: x_val2, training: False})
+        ema_mean_eval2 = ema_mean.eval()
+
+    sess.close()
+
+    assert_str = 'ema mean and var are not used correctly' + \
+                 'during testing with batch norm'
+    assert (not np.all(y_eval1 == np.zeros(input_shape))), assert_str
```
Comment (jrock08) [inline aiutils/tftools/test_batch_normalizer.py]: This depends on the initialization of the EMA, it's probably safe to assume the EMA is initialized to 0, but it *might* be worth mentioning that in a comment right above this assert so that what is being tested is more obvious.

Alternatively, it could be worth setting decay = 0 and then running once with x_val=N(u=1.5, sigma=.5), then we can say the y_eval1 ~= -1 and y_eval2 ~= 1.

File: aiutils/tftools/batch_normalizer.py [hunk 2]
```diff
@@ -95,3 +97,62 @@ def batch_normalize(input, phase_train, name):
...
+            self.output = tf.cond(training,
+                                  lambda: self.get_normalizer(input, True),
+                                  lambda: self.get_normalizer(input, False))
+
+    def get_normalizer(self, input, train_flag):
+        if train_flag:
+            ema_apply_op = self.ema.apply([self.mean, self.variance])
+            with tf.control_dependencies([ema_apply_op]):
+                self.output_training = tf.nn.batch_normalization(
+                    input, self.mean, self.variance, self.offset, self.scale,
```
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: How would you feel about moving the definition of self.mean and self.variance inside this, and then moving them to local scope instead of object scope?

Then the only thing at object scope would be offset, scale, ema, and output which I think makes more sense.

File: aiutils/tftools/batch_normalizer.py [hunk 3]
```diff
@@ -15,6 +15,8 @@
     optimizer = tf.group(update_assignments)
 """
 
+import pdb
+import numpy as np
```
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: pbd and numpy are unused?

**Comment by jrock08** [comment]
Looks good, also switch the batch_norm layer to use this impl and delete the old class.

File: aiutils/tftools/test_batch_normalizer.py (hunk 1 above)
Comment (daeyun) [inline aiutils/tftools/test_batch_normalizer.py]: The shadow variables are initialized to 0 for tf.Tensor and .initialized_value() for tf.Variable. (I just happen to have read it today) https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/moving_averages.py#L240

File: aiutils/tftools/test_batch_normalizer.py [hunk 4]
```diff
@@ -85,3 +85,86 @@ def test_batch_normalize_fcn(graph):
...
+        ema_mean_eval1 = ema_mean.eval()
+
+        y_eval2 = y.eval(feed_dict={x: x_val2, training: True})
+        ema_mean_eval2 = ema_mean.eval()
+
+    sess.close()
+
+    assert_str = 'batch mean and var are not used correctly' + \
+                 'during training with batch norm'
+    assert (np.all(y_eval1 == np.zeros(input_shape))), assert_str
```
Comment (daeyun) [inline aiutils/tftools/test_batch_normalizer.py]: Another way is
```python
assert np.allclose(y_eval1, 0)
```
or
```python
assert np.alltrue(y_eval1 == 0)
```

File: aiutils/tftools/test_batch_normalizer.py (hunk 1 above)
Comment (BigRedT) [inline aiutils/tftools/test_batch_normalizer.py]: yeah I had actually checked that the shadow variables were initialized to 0 while writing the test code.

File: aiutils/tftools/batch_normalizer.py (hunk 2 above)
Comment (BigRedT) [inline aiutils/tftools/batch_normalizer.py]: Yeah makes sense.

File: aiutils/tftools/batch_normalizer.py (hunk 3 above)
Comment (BigRedT) [inline aiutils/tftools/batch_normalizer.py]: I was using them for debugging, will get rid of them.

**Comment by BigRedT** [comment]
@jrock08 moved tests to test directory and cleaned up stuff. Please verify before I merge.

File: aiutils/tftools/layers.py [hunk 5]
```diff
@@ -1,8 +1,14 @@
 import numpy as np
 import tensorflow as tf
+from batch_normalizer import BatchNorm
 
 
-def full(input, out_dim, name, gain=np.sqrt(2), func=tf.nn.relu, reuse_vars=False):
```
Comment (jrock08) [inline aiutils/tftools/layers.py]: Is this autoformatting at work?

File: aiutils/tftools/layers.py [hunk 6]
```diff
@@ -93,43 +99,28 @@ def conv2d(input,
...
 
     Args:
-      input (tensor): Input to the layer.
-        Should have shape `[batch, in_dim]` or `[batch, in_height, in_width, in_dim]`.
-        Must be one of the following types: `float32`, `float64`.
-      name (str): Name used by the `tf.variable_scope`.
-      reuse_vars (bool): Determine whether the layer should reuse variables
-        or construct new ones.  Equivalent to setting reuse in variable_scope
+        input (tensor): Tensor to be batch normalized
+        training (bool tensor): Boolean tensor of shape []
```
Comment (jrock08) [inline aiutils/tftools/layers.py]: I'd recommend defaulting this to tf.Constant(True)

File: aiutils/tftools/layers.py [hunk 7]
```diff
@@ -93,43 +99,28 @@ def conv2d(input,
...
+        decay (float): Decay used for exponential moving average
+        epsilon (float): Small constant added to variance to prevent
+            division of the form 0/0
+        name (string): variable scope name
+        reuse_vars (bool): Value passed to reuse keyword argument of
```
Comment (jrock08) [inline aiutils/tftools/layers.py]: Drop reuse, it's not actually doing anything.

**Comment by jrock08** [comment]
this got tough to read due to the thing trying to smart diff, but I think it's good.

File: aiutils/tftools/batch_normalizer.py [hunk 8]
```diff
@@ -1,97 +1,71 @@
...
-model optimizer force it, e.g., by:
-
-  update_assignments = tf.group(bn1.get_assigner(),
-                                bn2.get_assigner())
-  with tf.control_dependencies([optimizer]):
-    optimizer = tf.group(update_assignments)
-"""
-
+import pdb
+import numpy as np
```
Comment (jrock08) [inline aiutils/tftools/batch_normalizer.py]: Unused imports
